crates/oro-node-semver/src/lib.rs

```diff
@@ -25,22 +26,116 @@ const MAX_SAFE_INTEGER: u64 = 900_719_925_474_099;
+        // Taken partially from nom.
+        let prefix = &self.input.as_bytes()[..self.offset];
+
+        // Count the number of newlines in the first `offset` bytes of input
+        let line_number = bytecount::count(prefix, b'\n');
+
+        // Find the line that includes the subslice:
+        // Find the *last* newline before the substring starts
+        let line_begin = prefix
+            .iter()
+            .rev()
+            .position(|&b| b == b'\n')
+            .map(|pos| self.offset - pos)
+            .unwrap_or(0);
+
+        // Find the full line after that newline
+        let line = self.input[line_begin..]
+            .lines()
+            .next()
+            .unwrap_or(&self.input[line_begin..])
+            .trim_end();
+
+        // The (1-indexed) column number is the offset of our substring into that line
+        let column_number = self.input[self.offset..].as_ptr() as usize - line.as_ptr() as usize;
+
+        (line_number, column_number)
+    }
+}
+
+#[derive(Debug, Error, Eq, PartialEq)]
+pub enum SemverErrorKind {
+    #[error("Semver string can't be longer than {} characters.", MAX_LENGTH)]
+    MaxLengthError,
+    #[error("Incomplete input to semver parser.")]
+    IncompleteInput,
+    #[error("Failed to parse an integer component of a semver string: {0}")]
+    ParseIntError(ParseIntError),
+    #[error("Integer component of semver string is larger than JavaScript's Number.MAX_SAFE_INTEGER: {0}")]
+    MaxIntError(u64),
+    #[error("Failed to parse {0} component of semver string.")]
+    Context(&'static str),
+    #[error("An unspecified error occurred.")]
+    Other,
+}
+
+#[derive(Debug)]
+struct SemverParseError<I> {
+    input: I,
+    context: Option<&'static str>,
+    kind: Option<SemverErrorKind>,
 }
 
 impl Diagnostic for SemverError {
-    fn code(&self) -> DiagnosticCode {
-        use SemverError::*;
-        match self {
-            ParseError { code, .. } => *code,
+    fn category(&self) -> DiagnosticCategory {
+        let (row, col) = self.location();
+        DiagnosticCategory::Parse {
+            input: self.input.clone(),
+            path: None,
+            row,
+            col,
         }
     }
+
+    fn subpath(&self) -> String {
+        "semver".into()
+    }
+
+    fn advice(&self) -> Option<String> {
+        todo!()
+    }
+}
+
+impl<I> ParseError<I> for SemverParseError<I> {
+    fn from_error_kind(input: I, _kind: nom::error::ErrorKind) -> Self {
+        Self {
+            input,
+            context: None,
+            kind: None,
+        }
+    }
+
+    fn append(_input: I, _kind: nom::error::ErrorKind, other: Self) -> Self {
+        other
+    }
+}
+
+impl<I> ContextError<I> for SemverParseError<I> {
+    fn add_context(_input: I, ctx: &'static str, mut other: Self) -> Self {
+        other.context = Some(ctx);
+        other
+    }
+}
+
+impl<'a> FromExternalError<&'a str, SemverParseError<&'a str>> for SemverParseError<&'a str> {
```

Comment:
Would you mind throwing a comment in explaining why this is needed?
When I was looking at the nom errors, I came across this trait. I'm consufesed why
we'd need to implement a conversion from SemverParseError -the external error- to SemverParseError -the nom error? 🤷

Reply:
`FromExternalError` is only needed for when you use `map_res`, and the error value returned by any `map_res` function basically needs a `FromExternalError` implementation in order to work. The reason for implementing it for `SemverParseError` is because we have a standalone parser that returns this, which I guess is kinda messy, but it is what it is.